src/coreclr/src/tools/Common/TypeSystem/Common/CastingHelper.cs

```diff
@@ -29,6 +102,24 @@ public static bool IsCompatibleWith(this TypeDesc thisType, TypeDesc otherType)
                 return false;
             }
 
+            // Managed pointers are compatible only if they are pointer-element-compatible-with as described in ECMA I.8.7.2
+            if (thisType.IsByRef && otherType.IsByRef)
+            {
+                return AreVerificationTypesEqual(thisType.GetParameterType(), otherType.GetParameterType());
+            }
+
+            // Unmanaged pointers are handled the same way as managed pointers
+            if (thisType.IsPointer && otherType.IsPointer)
+            {
+                return AreVerificationTypesEqual(thisType.GetParameterType(), otherType.GetParameterType());
+            }
+
+            // Function pointers are compatible only if they are method-signature-compatible-with as described in ECMA I.8.7.1
+            if (thisType.IsFunctionPointer && otherType.IsFunctionPointer)
+            {
+                return IsMethodSignatureCompatibleWith(thisType, otherType);
+            }
```

Comment:
We should add something along the lines of:

```csharp
if (otherType.IsByRef || otherType.IsPointer || otherType.IsFunctionPointer)
{
    return false;
}
```

Passing those to `CanCastTo` will cause an assert in `CanCastToInternal`. `CanCastTo` expects a boxable input. If we boxed a pointer, something is wrong.

Reply:
+1

Reply:
I think it should check not just otherType but also thisType. Once we've finished the checks written above, any of those sorts of types are a false result.

Reply:
For function pointers, will it take calling convention compatibility into account?

Reply:
Looks like https://github.com/dotnet/runtime/pull/37454/files#diff-ac8d470d9a85bb53d9a33af63b36fe45R87 and I just missed it my first pass